File: android/app/src/main/java/com/nxtend/team35/yubiboard/camera/CameraProfile.kt
package com.nxtend.team35.yubiboard.camera

import android.util.Size

enum class CameraProfile(val width: Int, val height: Int, val debugOnly: Boolean = false) {
    HD_720(1280, 720),
    QHD_540(960, 540),
    VGA_480(640, 480),
    FHD_1080(1920, 1080, debugOnly = true),
    ;

    val size: Size get() = Size(width, height)

    companion object {
        val productionOrder = listOf(HD_720, QHD_540, VGA_480)

        fun from(size: Size): CameraProfile = entries.firstOrNull {
            it.width == size.width && it.height == size.height
        } ?: HD_720
    }
}

data class CameraFrameInfo(
    val requestedProfile: CameraProfile,
    val actualWidth: Int,
    val actualHeight: Int,
    val rotationDegrees: Int,
    val cropLeft: Int,
    val cropTop: Int,
    val cropRight: Int,
    val cropBottom: Int,
)

File: ios/Screact/Camera/CameraProfile.swift
import AVFoundation

/// Port of Android CameraProfile. The production order mirrors Android (720p -> 540p -> 480p),
/// and — like Android's DEFAULT_BACK_CAMERA — Screact uses the rear camera.
enum CameraProfile: CaseIterable {
    case hd720
    case qhd540
    case vga480
    case fhd1080

    var width: Int {
        switch self {
        case .hd720: return 1280
        case .qhd540: return 960
        case .vga480: return 640
        case .fhd1080: return 1920
        }
    }

    var height: Int {
        switch self {
        case .hd720: return 720
        case .qhd540: return 540
        case .vga480: return 480
        case .fhd1080: return 1080
        }
    }

    var debugOnly: Bool { self == .fhd1080 }

    /// The closest AVCaptureSession preset for this profile.
    var sessionPreset: AVCaptureSession.Preset {
        switch self {
        case .hd720: return .hd1280x720
        case .qhd540: return .iFrame960x540
        case .vga480: return .vga640x480
        case .fhd1080: return .hd1920x1080
        }
    }

    static let productionOrder: [CameraProfile] = [.hd720, .qhd540, .vga480]

    static func from(width: Int, height: Int) -> CameraProfile {
        allCases.first { $0.width == width && $0.height == height } ?? .hd720
    }
}

struct CameraFrameInfo {
    let requestedProfile: CameraProfile
    let actualWidth: Int
    let actualHeight: Int
    let rotationDegrees: Int
}
